core/market/decentralized/tests/test_negotiations.rs

```diff
@@ -140,3 +131,447 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
+        .counter_proposal(&offer_id, &proposal0.get_proposal_id()?, &proposal1)
+        .await;
+
+    assert!(result.is_err());
+    match result.err().unwrap() {
+        ProposalError::Unsubscribed(id) => assert_eq!(id, offer_id),
+        _ => panic!("Expected ProposalError::Unsubscribed."),
+    }
+
+    Ok(())
+}
+
+/// Requestor tries to counter initial Proposal, for which Offer was unsubscribed on remote Node.
+/// Negotiation attempt should be rejected by Provider Node.
+#[cfg_attr(not(feature = "market-test-suite"), ignore)]
+#[actix_rt::test]
+async fn test_counter_initial_unsubscribed_remote_offer() -> Result<(), anyhow::Error> {
+    let network = MarketsNetwork::new("test_counter_initial_unsubscribed_remote_offer")
+        .await
+        .add_market_instance("Node-1")
+        .await?
+        .add_market_instance("Node-2")
+        .await?;
+
+    let market1 = network.get_market("Node-1");
+    let market2 = network.get_market("Node-2");
+
+    let identity1 = network.get_default_id("Node-1");
+    let identity2 = network.get_default_id("Node-2");
+
+    let demand_id = market1
+        .subscribe_demand(&sample_demand(), &identity1)
+        .await?;
+    let offer_id = market2.subscribe_offer(&sample_offer(), &identity2).await?;
+
+    let proposal0 = requestor::query_proposal(&market1, &demand_id, 1).await?;
+
+    // When we will counter this Proposal, Provider will have it already unsubscribed.
+    market2.unsubscribe_offer(&offer_id, &identity2).await?;
+
+    let proposal1 = proposal0.counter_demand(sample_demand())?;
+    let result = market1
+        .requestor_engine
+        .counter_proposal(&demand_id, &proposal0.get_proposal_id()?, &proposal1)
+        .await;
+
+    assert!(result.is_err());
+    match result.err().unwrap() {
+        ProposalError::FailedSendProposal(..) => (),
+        _ => panic!("Expected ProposalError::FailedSendProposal."),
+    }
+
+    Ok(())
+}
+
+/// Requestor tries to counter draft Proposal, for which Offer was unsubscribed on remote Node.
+/// Negotiation attempt should be rejected by Provider Node.
+#[cfg_attr(not(feature = "market-test-suite"), ignore)]
+#[actix_rt::test]
+async fn test_counter_draft_unsubscribed_remote_offer() -> Result<(), anyhow::Error> {
+    let network = MarketsNetwork::new("test_counter_initial_unsubscribed_remote_offer")
+        .await
+        .add_market_instance("Node-1")
+        .await?
+        .add_market_instance("Node-2")
+        .await?;
+
+    let NegotiationHelper {
+        proposal_id: proposal0_id,
+        proposal: proposal0,
+        offer_id,
+        demand_id,
+    } = exchange_draft_proposals(&network, "Node-1", "Node-2").await?;
+
+    let market1 = network.get_market("Node-1");
+    let market2 = network.get_market("Node-2");
+    let identity2 = network.get_default_id("Node-2");
+
+    // When we will counter this Proposal, Provider will have it already unsubscribed.
+    market2.unsubscribe_offer(&offer_id, &identity2).await?;
```

Comment:
again, how do we behave when Offer unsubscription (on Provider) will occur before Requestor queries for Proposals?

Reply:
#510